### DISCUSSION TITLE
Cannot use strftime with constructed format string

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#9998] [Ideas]
### What happens?

The following query raises the error `Error: near line 4: Invalid Input Error: strftime format must be a constant`

```sql
with
  specs as (select unnest(split('a;A;w;d;-d;b;B;m;-m;y;-y;Y;H;-H;I;-I;p;M;-M;S;-S;g;f;n;z;Z;j;-j;U;W;c;x;X;%',';')) as spec)
, fmts as (select spec, '%'||spec as fmt from specs)
select spec,'strftime(current_date,'''||fmt||''')' as call , strftime(current_date,fmt) as val
from fmts
```

A similar function in bigquery works without any errors!

```sql
with
  specs as (select spec from unnest(split('a;A;w;d;-d;b;B;m;-m;y;-y;Y;H;-H;I;-I;p;M;-M;S;-S;g;f;n;z;Z;j;-j;U;W;c;x;X;%',';')) as spec)
, fmts as (select spec, '%'||spec as fmt from specs)
select spec,"strftime(current_date,'"||fmt||"')" as call , format_date(fmt,current_date) as val
from fmts
```

### To Reproduce

```sql
with
  specs as (select unnest(split('a;A;w;d;-d;b;B;m;-m;y;-y;Y;H;-H;I;-I;p;M;-M;S;-S;g;f;n;z;Z;j;-j;U;W;c;x;X;%',';')) as spec)
, fmts as (select spec, '%'||spec as fmt from specs)
select spec,'strftime(current_date,'''||fmt||''')' as call , strftime(current_date,fmt) as val
from fmts
```

### OS:

Darwin 22.6.0 arm64

### DuckDB Version:

v0.9.2 3c695d7ba9

### DuckDB Client:

duckdb cli

### Full Name:

Vijay Lulla

### Affiliation:

Sweetwater

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Mytherin** [comment]
Thanks for the report, this is not a bug but an expected restriction. Removing the restriction is a feature request and hence belongs in discussions.

**Reply by vlulla** [reply]
Oops sorry! Yes, this is indeed more appropriate for a discussion.

**Reply by vlulla** [reply]
Is there any interest in allowing the `format` argument to be a string expression instead of a string literal? If so, I'm willing to try to implement this...but i will need a little bit of guidance (about where to look and how to test the implementation). 

I have looked at `src/core_functions/scalar/date/strftime.cpp` and `src/execution/expression_executor.cpp`. Are these the correct places?

**Reply by Tishj** [reply]
By string expression you mean not evaluated at bind time?
I think you're on the right path, I don't know if this is something we're interested in, but that's not my call.

What happens currently is in `StrfTimeBindFunction` we evaluate the format string and save it in the `StrfTimeBindData`.
This is then provided to the `StrfTimeFunctionTimestamp` function, which uses it to format the input chunk accordingly.

Your version would probably need to do this evaluation of the format string for every tuple, using the same parameter that we use in the bind data

So you'd add a new overload, changing the existing one to use `LogicalType::STRING_LITERAL` for the `format` type, so the binder can differentiate between the two function overloads.
(in `StrfTimeFun::GetFunctions()`)

Then for your new function write a Bind function, which basically just returns a basic `FunctionData`, storing nothing extra

And finally add a new Function for your function, which evaluates the expression (same as done in the regular `StrfTimeBindFunction` function, then loop over your produced result Vector of strings, for every iteration turn it into a `StrfTimeFormat` and do validation.
Then use the format to convert the input string: `format.ConvertTimestampVector(args.data[REVERSED ? 1 : 0], result, args.size())`
